signingworker/worker.py

```diff
@@ -0,0 +1,187 @@
+            json.dump(signing_manifest, f, indent=2, sort_keys=True)
+        log.debug("Uploading manifest for t: %s, r: %s", task_id, run_id)
+        self.create_artifact(task_id, run_id, "public/env/manifest.json",
+                             manifest_file, "application/json")
+
+    def download_and_sign_file(self, task_id, run_id, url, checksum):
+        work_dir = tempfile.mkdtemp()
+        # TODO: better parsing
+        filename = urlparse.urlsplit(url).path.split("/")[-1]
+        abs_filename = os.path.join(work_dir, filename)
+        r = requests.get(url)
+        r.raise_for_status()
+        with open(abs_filename, 'wb') as fd:
+            for chunk in r.iter_content(4096):
+                fd.write(chunk)
+        got_checksum = get_hash(abs_filename)
+        if not got_checksum == checksum:
+            log.debug("Checksum mismatch, cleaning up...")
+            shutil.rmtree(work_dir)
+            raise ChecksumMismatchError("Expected {}, got {} for {}".format(
+                checksum, got_checksum, url
+            ))
+        self.sign_file(work_dir, filename)
+        self.create_artifact(task_id, run_id, "public/env/%s" % filename,
+                             abs_filename)
+        return abs_filename
+
+    def create_artifact(self, task_id, run_id, dest, abs_filename,
+                        content_type="application/octet-stream"):
+        log.debug("Uploading artifact %s (t: %s, r: %s) from %s (%s)", dest,
+                  task_id, run_id, abs_filename, content_type)
+        # TODO: better expires
+        res = self.tc_queue.createArtifact(
+            task_id, run_id, dest,
+            {
+                "storageType": "s3",
+                "contentType": content_type,
+                "expires": arrow.now().replace(weeks=2).isoformat()
+            }
+        )
+        log.debug("Got %s", res)
+        put_url = res["putUrl"]
+        log.debug("Uploading to %s", put_url)
+        taskcluster.utils.putFile(abs_filename, put_url, content_type)
+
+    @redo.retriable(attempts=10, sleeptime=5, max_sleeptime=30)
+    def get_token(self, output_file):
+        token = None
+        data = {"slave_ip": my_ip, "duration": 5 * 60}
+        random.shuffle(self.signing_servers)
+        for server in self.signing_servers:
+            log.debug("getting token from %s", server)
+            # TODO: Figure out how to deal with certs not matching hostname,
+            #  error: https://gist.github.com/rail/cbacf2d297decb68affa
+            r = requests.post("https://{}/token".format(server), data=data,
+                              auth=tuple(self.signing_server_auth),
+                              verify=False)
+            r.raise_for_status()
+            if r.content:
+                token = r.content
+                break
+        if not token:
+            raise SigningServerError("Cannot retrieve signing token")
+        with open(output_file, "wb") as f:
+            f.write(token)
```

Comment:
The flow here might be better if you wrote it as a for...else block. You could move the writing of the token into the loop, and then change "if not token" to "else". I'm happy either way, just a thought.